include/nbl/builtin/hlsl/workgroup/scratch.hlsl

```diff
@@ -0,0 +1,116 @@
+// Copyright (C) 2023 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+#ifndef _NBL_BUILTIN_HLSL_SUBGROUP_SCRATCH_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SUBGROUP_SCRATCH_INCLUDED_
+
+#include "nbl/builtin/hlsl/glsl_compat/basic.hlsl"
+#include "nbl/builtin/hlsl/glsl_compat/subgroup_basic.hlsl"
+#include "nbl/builtin/hlsl/workgroup/basic.hlsl"
+#include "nbl/builtin/hlsl/subgroup/basic.hlsl"
+
+/**
+ * The clearIndex for each invocation is the first 8 indices for each set of subgroup::Size().
+ * Example for subgroup::Size() == 32:
+ *	invocationId -> clearIndex
+ *	1 -> 1
+ *	7 -> 7
+ *	8 -> 32
+ *	15 -> 39
+ *	16 -> 64
+ *	24 -> 96
+ */
+inline uint scratchInitializeClearIndex(uint invocationId, uint halfSubgroupMask)
+{
+	return ((((invocationId)&(~halfSubgroupMask))<<2u)|((invocationId)&halfSubgroupMask));
+}
+
+namespace nbl 
+{
+namespace hlsl
+{
+namespace subgroup
+{
+	struct ScratchOffsetsAndMasks
+	{
+		/**
+		 * Struct creator with default values for subgroup portability arithmetic operations.
+		 * For each subgroup inside the workgroup, the memory range is SubgroupSize + HalfSubgroupSize
```

Comment:
you now only need 1 shared memory variable per subgroup just to "store the head" reduction value